harness/determined/keras/_tf_keras_trial.py

```diff
@@ -356,36 +324,135 @@ def from_native(
     ) -> None:
-        (
-            context.model,
-            compile_args.arguments["optimizer"],
-        ) = keras._get_multi_gpu_model_and_optimizer(
-            pre_compiled_model=context.model,
-            optimizer=compile_args.arguments["optimizer"],
-            env=env,
-            hvd_config=hvd_config,
+        super().__init__(*args, **kwargs)
+
+        self.model = model
+        self.session = session
+
+        self._train_input_manager, self._validation_input_manager = keras._init_input_managers(
+            context=self.context, train_config=train_config
         )
 
-        if hvd_config.use and version.parse("2.0.0") <= version.parse(
-            tf.__version__
-        ) < version.parse("2.2.0"):
-            logging.info(
-                "Calling `model.compile(...)` with `experimental_run_tf_function=False` to ensure "
-                "TensorFlow calls `optimizer.get_gradients()` to compute gradients."
-            )
+        self._configure_callbacks(train_config.callbacks, train_config.verbose)
 
-            context.model.compile(
-                *compile_args.args, **compile_args.kwargs, experimental_run_tf_function=False
-            )
-        else:
-            context.model.compile(*compile_args.args, **compile_args.kwargs)
+        # If a load path is provided, load weights and restore the data location.
+        self._load()
+
+        self.train_response_func = None  # type: Optional[workload.ResponseFunc]
+        self.train_workload_metrics = []  # type: List[Dict[str, Any]]
+        self.train_workload_batches = 0
+        self.train_workload_inputs = 0
+        self.train_workload_len = 0
+        self.test_inputs = 0
+
+    def _configure_callbacks(self, user_callbacks: Optional[List], verbose: int) -> None:
+        """
+        If we pass a callbacks parameter to model.fit() or model.evaluate() which is a
+        pre-constructed CallbackList, Keras will not alter it.  We can use this property to
+        configure the exact callback order that we want in our system.
+
+        The implementation is based closely on from the real
+        tf.keras.callbacks.configure_callbacks(), with the following differences:
+
+          - We always assume we have the original Callbacks list.
+          - We prepend and append additional Determined and Horovod callbacks
+          - We create a det.keras.CallbackList instead of the normal tf.keras one.
+        """
+
+        callbacks = user_callbacks or []
+        assert isinstance(callbacks, list), (
+            "the callbacks parameter of model.fit() or model.train() must be a list of Callbacks, "
+            f"not {type(callbacks)}"
+        )
+
+        if self.env.experiment_config.get_records_per_epoch() is None:
+            for cb in callbacks:
+                if not isinstance(cb, keras.callbacks.Callback):
+                    continue
+                if util.is_overridden(cb.on_epoch_end, tf.keras.callbacks.Callback):
```

Comment:
question: should we check `on_epoch_begin` as well?

Reply:
cool, leaving it as-is.